### PR TITLE
Improve abstract InternalRuntime to be more flexible for recovery functionality

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
Improve abstract InternalRuntime to be more flexible for recovery functionality. It includes:

- Add an ability to store status by an implementation of InternalRuntime. It allows infrastructure provide an actual status after runtime recovery. Also, it adds an ability to deny stopping of STARTING workspaces if it is not implemented like in Kubernetes/OpenShift infrastructures.
Unfortunately, denying stop of STARTING workspace will not work correctly. Here is a registered bug for that https://github.com/eclipse/che/issues/9354.

- Add an ability to throw InfrastructureException by InternalRuntimes#getMachines method. It is needed because Runtime may store machines information in storage that may throw some exceptions (like database as storage may throw connection related exceptions).

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/5919
It's needed for OpenShift recovery functionality changes https://github.com/eclipse/che/pull/9301

### CODE REVIEW

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java [hunk 1]
```diff
@@ -144,16 +144,25 @@ public void validate(Environment environment)
    *
    * @param workspace the workspace to inject runtime into
    */
-  public void injectRuntime(WorkspaceImpl workspace) {
+  public void injectRuntime(WorkspaceImpl workspace) throws ServerException {
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java:147]: I'd rather use InfrastructureException. We have discussed that ServerException if rather REST exception.

**Review by akorneta — APPROVED** [review]

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/InternalRuntime.java [hunk 2]
```diff
@@ -86,9 +112,21 @@ public String getOwner() {
...
-   * <p>Implementation should not return null
+   * <p>Implementation must not return null
+   *
+   * @throws InfrastructureException when any error occurs
+   */
+  protected abstract Map<String, ? extends Machine> getInternalMachines()
+      throws InfrastructureException;
+
+  /**
+   * Returns workspace status.
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/InternalRuntime.java]: Runtime status or workspace status?

**Review by garagatyi — APPROVED** [review]
LGTM

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java (hunk 1 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java:147]: I agree that `ServerException` is REST exception, but `InfrastructureException` is an exception that is supposed to be thrown by Infrastructure. As you can see, now `WorkspaceRuntimes` throw only REST exceptions (Conflict, NotFound, Server).
I'd say that WorkspaceRuntimes should not throw REST exception but and Infrastructure exception isn't the right one here too. I guess there is an issue for reworking Che Master exception approach and this issue could be solved there.
@garagatyi Does it makes sense?

**Comment by sleshchenko** [comment]
ci-test

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java (hunk 1 above)
Comment (garagatyi) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java:147]: Yep, you are completely right. I missed that.

**Comment by sleshchenko** [comment]
ci-test

**Review by skabashnyuk — APPROVED** [review]

**Comment by sleshchenko** [comment]
ci-test

**Pull request merged**